specs/agents/tracing-spans-drop-fast-exit.md

```diff
@@ -0,0 +1,82 @@
+Therefore, agents SHOULD implement the strategy laid out in this section to let users choose the level of detail/cost tradeoff that makes sense for them.
+If an agent implements this strategy, it MUST also implement [Collecting statistics about dropped spans](tracing-spans-dropped-stats.md).
+
+## `exit_span_min_duration` configuration
+
+Sets the minimum duration of exit spans.
+Exit spans that execute faster than this threshold are attempted to be discarded.
+
+The attempt fails if they lead up to a span that can't be discarded.
+Spans that propagate the trace context to downstream services,
+such as outgoing HTTP requests,
+can't be discarded.
+However, external calls that don't propagate context,
+such as calls to a database, can be discarded using this threshold.
+
+Additionally, spans that lead to an error can't be discarded.
+
+|                |            |
+|----------------|------------|
+| Type           | `duration` |
+| Default        | `1ms`      |
+| Central config | `true`     |
+
+TODO: should we introduce µs granularity for this config option?
+Adding `us` to all `duration`-typed options would create compatibility issues.
+So we probably want to support `us` for this option only.
+
+## Interplay with span compression
+
+If an agent implements [span compression](tracing-spans-compress.md),
+the limit applies to the [composite span](tracing-spans-compress.md#composite-span).
+
+For example, if 10 Redis calls are compressed into a single composite span whose total duration is lower than `exit_span_min_duration`,
+it will be dropped.
+If, on the other hand, the individual Redis calls are below the threshold,
+but the sum of their durations is above it, the composite span will not be dropped.
+
+## Limitations
+
+The limitations are based on the premise that the `parent_id` of each span and transaction that's stored in Elasticsearch
+should point to another valid transaction or span that's present in the Elasticsearch index.
+
+A span that refers to a missing span via is `parent_id` is also known as an "orphaned span".
+
+### Spans that propagate context to downstream services can't be discarded
+
+We only know whether to discard after the call has ended.
+At that point,
+the trace has already continued on the downstream service.
+Discarding the span for the external request would orphan the transaction of the downstream call.
+
+Propagating the trace context to downstream services is also known as out-of-process context propagation.
+
+## Implementation
+
+### `discardable` flag
+
+Spans store an additional `discardable` flag in order to determine whether a span can be discarded.
+The default value is `true` for [exit spans](tracing-spans.md#exit-spans) and `false` for any other span.
+
+According to the [limitations](#Limitations),
+there are certain situations where the `discardable` flag of a span is set to `false`:
+- the span has `outcome` (i.e., `outcome` is present and it's not `null`) `outcome` is not `success`.
```

Comment:
```suggestion
- the span's `outcome` field is set to anything other than `success`.
```

Reply:
Do you mean that a span with `outcome` field (or `outcome` with `null` value) should not be compressed?

Reply:
Yes, that's what I meant. Agents shouldn't end spans without setting an outcome to them, it is allowed to be omitted for backward compatibility reason IIANM. So, assuming these are most likely custom user spans, I think we should default to not compressing them.
In general, since the main goal of this initiative is to reduce the amount of spans with minimal information loss, I think there should be rare cases of spans without outcome and therefore the default should be to not drop them. The fix for the inability to compress should be setting the outcome.
But I don't mind if we start with defaulting to compress.